### PR TITLE
Allow resumption of indexing jobs.

### LINKED ISSUE #62: Rebuild index leaks memory
When executing the index rebuild command on ~ 10.000 Contacts:
app/console massive:search:index:rebuild

a OutOfMemoryException accurs:
[Symfony\Component\Debug\Exception\OutOfMemoryException]                                  
Error: Allowed memory size of 1073741824 bytes exhausted (tried to allocate 65552 bytes)

### PR DESCRIPTION
**Pull request by dantleech**
This PR introduces the `RebuildResumeManager` which allows listeners to register the progress they have made in order that they can later resume it.

It there is progress registered, the `RebuildIndexCommand` will ask the user if they wish to resume.

Fixes #62

### CODE REVIEW

File: Command/IndexRebuildCommand.php [hunk 1]
```diff
@@ -39,20 +41,37 @@ public function configure()
...
         $purge = $input->getOption('purge');
         $filter = $input->getOption('filter');
-        $debug = $input->getOption('debug');
-
-        if (!$debug) {
-            $this->getContainer()
-                ->get('doctrine.orm.default_entity_manager')
-                ->getConnection()
-                ->getConfiguration()
-                ->setSQLLogger(null);
```
Comment (danrot) [inline Command/IndexRebuildCommand.php]: Why did you remove that? This code reduced memory usage by a lot.

File: Search/Command/RebuildResumeManager.php [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    public function getCheckpoints()
+    {
+        $filename = $this->getCheckpointFile();
+        $data = array();
+
+        if (file_exists($filename)) {
+            $data = unserialize(file_get_contents($filename));
+        }
+
+        if (!$data || !is_array($data)) {
```
Comment (danrot) [inline Search/Command/RebuildResumeManager.php]: Wouldn't only `!is_array($data)` do exactly the same?

File: Search/Command/RebuildResumeManager.php [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+ * This is basically a key/value store.
+ */
+class RebuildResumeManager
+{
+    public function getCheckpointFile()
+    {
+        return sys_get_temp_dir() . '/' . str_replace('\\', '_', __CLASS__);
+    }
+
+    public function checkpoint($name, $value)
```
Comment (danrot) [inline Search/Command/RebuildResumeManager.php]: Where is this actually called?

**Comment by danrot** [comment]
We also have now a changelog in this bundle :wink:

File: Search/Command/RebuildResumeManager.php (hunk 3 above)
Comment (dantleech) [inline Search/Command/RebuildResumeManager.php]: It is called from the listener. It should be renamed to `setCheckpoint` or similar.

File: Search/Command/RebuildResumeManager.php (hunk 2 above)
Comment (dantleech) [inline Search/Command/RebuildResumeManager.php]: i think so.

File: Command/IndexRebuildCommand.php [hunk 4]
```diff
@@ -52,7 +60,33 @@ public function execute(InputInterface $input, OutputInterface $output)
                 ->setSQLLogger(null);
```
Comment (dantleech) [inline Command/IndexRebuildCommand.php]: @danrot still not sure that we should be doing this here -- it must be possible to disable DBAL logging when `%kernel.debug%` is false.

File: Command/IndexRebuildCommand.php [hunk 5]
```diff
@@ -52,7 +60,33 @@ public function execute(InputInterface $input, OutputInterface $output)
...
         }
 
+        $checkpoints = $resumeManager->getCheckpoints();
+
+        $questionHelper = new QuestionHelper();
```
Comment (dantleech) [inline Command/IndexRebuildCommand.php]: Add warning if in `dev` mode and recommend starting again in `prod`.

File: Search/ReIndex/ExceptionCollection.php [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Massive\Bundle\SearchBundle\Search\ReIndex;
+
+/**
+ * Collection of errors which occured during the reindexing processes.
+ */
+class ErrorCollection
```
Comment (dantleech) [inline Search/ReIndex/ExceptionCollection.php]: Remove

File: Search/Event/IndexRebuildEvent.php [hunk 7]
```diff
@@ -21,6 +22,10 @@
  */
 class IndexRebuildEvent extends Event
 {
+    private $filter;
+    private $purge;
+    private $output;
```
Comment (dantleech) [inline Search/Event/IndexRebuildEvent.php]: `@var` doc

File: Search/ReIndex/ResumeManager.php [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+        $data = $this->getCheckpoints();
+
+        if (isset($data[$name])) {
+            return $data[$name];
+        }
+
+        return $default;
+    }
+
+    public function removeCheckpoint($name)
```
Comment (dantleech) [inline Search/ReIndex/ResumeManager.php]: Doc methods.

File: Tests/Unit/Search/ReIndex/ResumeManagerTest.php [hunk 9]
```diff
@@ -0,0 +1,89 @@
+<?php
+
+namespace Massive\Bundle\SearchBundle\Tests\Unit\Search\Command;
```
Comment (dantleech) [inline Tests/Unit/Search/ReIndex/ResumeManagerTest.php]: Add styleci

File: Tests/Unit/Search/ReIndex/ResumeManagerTest.php [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+use Massive\Bundle\SearchBundle\Search\ReIndex\ResumeManager;
+
+class ResumeManagerTest extends \PHPUnit_Framework_TestCase
+{
+    private $manager;
```
Comment (dantleech) [inline Tests/Unit/Search/ReIndex/ResumeManagerTest.php]: doc

File: Command/StatusCommand.php [hunk 11]
```diff
@@ -26,7 +26,7 @@ public function configure()
     {
         $this->setName('massive:search:status');
         $this->setDescription('Return the status of the configured search engine');
-        $this->setHelp(<<<EOT
+        $this->setHelp(<<<'EOT'
```
Comment (dantleech) [inline Command/StatusCommand.php]: StyleCI does this, I have no idea why.

Comment (dantleech) [inline Command/StatusCommand.php]: It is a nowdoc and contrary to a heredoc nothing inside it is parsed.

File: Tests/Unit/Search/ReIndex/ResumeManagerTest.php [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+    /**
+     * @var ResumeManager
+     */
+    private $manager;
+
+    public function setUp()
+    {
+        $this->manager = new ResumeManager();
+        $this->manager->purgeCheckpoints();
+    }
```
Comment (dantleech) [inline Tests/Unit/Search/ReIndex/ResumeManagerTest.php]: Add teardown

File: Command/IndexRebuildCommand.php (hunk 4 above)
Comment (dantleech) [inline Command/IndexRebuildCommand.php]: Am leaving this here unchanged however.

**Comment by dantleech** [comment]
@danrot @wachterjohannes @chirimoya ready for review.

File: Command/IndexRebuildCommand.php [hunk 13]
```diff
@@ -14,12 +14,18 @@
...
 use Symfony\Component\Console\Input\InputInterface;
 use Symfony\Component\Console\Input\InputOption;
 use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Console\Question\ConfirmationQuestion;
+use Symfony\Component\Console\Question\Question;
 
 /**
  * Comand to build (or rebuild) the search index.
+ *
+ * TODO: Rename this to ReIndexCommand
```
Comment (danrot) [inline Command/IndexRebuildCommand.php]: How about renaming it now, and creating this class again, decorating the renamed class, and mark it as deprecated?

File: Command/IndexRebuildCommand.php [hunk 14]
```diff
@@ -39,7 +45,28 @@ public function configure()
...
+        $formatterHelper = new FormatterHelper();
+        $questionHelper = new QuestionHelper();
+
+        $container = $this->getContainer();
+        $startTime = microtime(true);
+        $eventDispatcher = $container->get('event_dispatcher');
+        $resumeManager = $container->get('massive_search.reindex.resume_manager');
+        $env = $container->getParameter('kernel.environment');
+
+        if ($env !== 'prod') {
```
Comment (danrot) [inline Command/IndexRebuildCommand.php]: Would also add an option to disable this warning, so that people using this command and being aware of that don't have their output cluttered.

File: Command/IndexRebuildCommand.php (hunk 4 above)
Comment (danrot) [inline Command/IndexRebuildCommand.php]: Well, if you find a way, I've got nothing against it :wink:

File: Search/ReIndex/ResumeManager.php [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+{
+    /**
+     * Store a checkpoint. This would typically be an offset
+     * from where the reindexer can subsequently resume its
+     * task.
+     *
+     * @param mixed $name
+     * @param mixed $value
+     */
+    public function setCheckpoint($name, $value)
```
Comment (danrot) [inline Search/ReIndex/ResumeManager.php]: Where is this method actually called?

File: Search/ReIndex/ResumeManager.php [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+
+/**
+ * Allow the storage of "checkpoints" which indicate how far
+ * a reindexing process has progressed in the case that it is
+ * interupted (f.e. by an error) with the aim of allowing the
+ * reindexing process to resume from where it left off.
+ *
+ * This is basically a key/value store.
+ */
+class ResumeManager
```
Comment (danrot) [inline Search/ReIndex/ResumeManager.php]: I would introduce an interface here.

File: Search/ReIndex/ResumeManager.php (hunk 15 above)
Comment (dantleech) [inline Search/ReIndex/ResumeManager.php]: From sulu atm.

Comment (danrot) [inline Search/ReIndex/ResumeManager.php]: Why does Sulu have to know about this?

Comment (dantleech) [inline Search/ReIndex/ResumeManager.php]: The Sulu Structure Reindexer has this class injected into it.

**Comment by dantleech** [comment]
Updated. Renamed command, also made the command a service.

File: Search/ReIndex/ResumeManager.php (hunk 15 above)
Comment (danrot) [inline Search/ReIndex/ResumeManager.php]: In which PR? Nothing is linked here. And that means it only works for pages, but not for anything else?

**Comment by danrot** [comment]
Tests failing ...

**Comment by dantleech** [comment]
Updated.

File: Search/ReIndex/ResumeManager.php (hunk 15 above)
Comment (dantleech) [inline Search/ReIndex/ResumeManager.php]: It is not implemented on the ORM

Comment (danrot) [inline Search/ReIndex/ResumeManager.php]: But that's not a good thing... I.e. we have projects where loads of assets are uploaded, so this functionality would be even more important for the assets in these projects. Would it be hard to implement the resume functionality directly in here? Would also make more sense IMO.

Comment (dantleech) [inline Search/ReIndex/ResumeManager.php]: We could inject it into the ORM listener or we could refactor
everything, which would be great. But I do not want to spend the next
week on this.

File: Command/ReindexCommand.php [hunk 17]
```diff
@@ -0,0 +1,234 @@
...
+use Symfony\Component\Console\Helper\ProgressBar;
+use Symfony\Component\Console\Helper\QuestionHelper;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Console\Question\ConfirmationQuestion;
+use Symfony\Component\Console\Question\Question;
+
+/**
+ * Comand to build (or rebuild) the search index.
```
Comment (danrot) [inline Command/ReindexCommand.php]: `Command`

File: DependencyInjection/Compiler/ReindexProviderPass.php [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+ */
+
+namespace Massive\Bundle\SearchBundle\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * Compiler pass to register metadata providers.
```
Comment (danrot) [inline DependencyInjection/Compiler/ReindexProviderPass.php]: `reindex providers`

File: Resources/config/search.xml [hunk 19]
```diff
@@ -80,5 +80,12 @@
 
             <tag name="kernel.event_listener" event="%massive_search.events.deindex%" method="onDeindex"/>
         </service>
+
+        <service id="massive_search.reindex.resume_manager" class="Massive\Bundle\SearchBundle\Search\Reindex\ResumeManager"/>
+
+        <!-- Reindex -->
+        <service id="massive_search.reindex.provider_registry" class="Massive\Bundle\SearchBundle\Search\Reindex\ReindexProviderRegistry" />
+        <service id="massive_search.reindex.resume_manager" class="Massive\Bundle\SearchBundle\Search\Reindex\ResumeManager"/>
```
Comment (danrot) [inline Resources/config/search.xml]: Double definition of `ResumeManager`

File: Resources/docs/extending.rst [hunk 20]
```diff
@@ -95,6 +95,20 @@ the converter to the system you simply add a tag to your custom service.
         <tag type="massive_search.converter" from="foo" />
     </service>
 
+Reindex Providers
+-----------------
+
+When you implement a new driver you will most likely want to be able to
+*re-index* objects which fall within the scope of this driver. In order to do
+this you must create a class implementing ``ReindexProviderInterface`` and
+add it to your service configutation with the ``massive_search.reindex.provider`` tag::
```
Comment (danrot) [inline Resources/docs/extending.rst]: Only one colon at the end

**Comment by dantleech** [comment]
Updated.

File: Resources/config/search.xml (hunk 19 above)
Comment (danrot) [inline Resources/config/search.xml]: @dantleech There are still two definitions for the resume manager.

Comment (dantleech) [inline Resources/config/search.xml]: Fixed.
